identity_presentation/src/iosMain/kotlin/net/thechance/mena/identity/presentation/util/permissions/CameraPermission.kt

```diff
@@ -0,0 +1,85 @@
+import net.thechance.mena.identity.presentation.util.permissionHandler.PermissionState
+import platform.AVFoundation.*
+import kotlin.coroutines.resume
+import kotlin.coroutines.resumeWithException
+
+internal class CameraPermission : PermissionController {
+
+    override fun getPermissionState(): PermissionState {
+        val status = AVCaptureDevice.authorizationStatusForMediaType(AVMediaTypeVideo)
+        return status.toPermissionState()
+    }
+
+    override fun openSettingPage() {
+        openAppSettingsPage()
+    }
+
+    override suspend fun requestPermission() {
+        val beforeStatus = AVCaptureDevice.authorizationStatusForMediaType(AVMediaTypeVideo)
+
+        if (beforeStatus == AVAuthorizationStatusAuthorized) return
+
+        return suspendCancellableCoroutine { cont ->
+            AVCaptureDevice.requestAccessForMediaType(AVMediaTypeVideo) { granted ->
+                if (cont.isCancelled) return@requestAccessForMediaType
+
+                val afterStatus = AVCaptureDevice.authorizationStatusForMediaType(AVMediaTypeVideo)
+                handleAuthorizationChange(beforeStatus, afterStatus, cont, granted)
+                }
+            }
+        }
+    }
+
+    private fun AVAuthorizationStatus.toPermissionState(): PermissionState {
+        return when (this) {
+            AVAuthorizationStatusAuthorized -> PermissionState.GRANTED
+            AVAuthorizationStatusNotDetermined -> PermissionState.NOT_DETERMINED
+            AVAuthorizationStatusDenied -> PermissionState.DENIED
+            AVAuthorizationStatusRestricted -> PermissionState.DENIED_PERMANENTLY
+            else -> PermissionState.NOT_DETERMINED
+        }
+    }
+
+    private fun handleAuthorizationChange(
+        beforeStatus: AVAuthorizationStatus,
+        afterStatus: AVAuthorizationStatus,
+        cont: CancellableContinuation<Unit>,
+        granted: Boolean
+    ){
+        when {
+            granted || afterStatus == AVAuthorizationStatusAuthorized -> {
+                cont.resume(Unit)
+            }
+
+            beforeStatus == AVAuthorizationStatusNotDetermined &&
+                    afterStatus == AVAuthorizationStatusDenied -> {
+                cont.resumeWithException(PermissionDeniedException())
+            }
+
+            beforeStatus == AVAuthorizationStatusDenied &&
+                    afterStatus == AVAuthorizationStatusDenied -> {
+                cont.resumeWithException(PermissionDeniedPermanentlyException())
+            }
+
+            afterStatus == AVAuthorizationStatusRestricted -> {
+                cont.resumeWithException(PermissionDeniedPermanentlyException())
+            }
+
+            afterStatus == AVAuthorizationStatusNotDetermined -> {
+                cont.resumeWithException(PermissionNotDeterminedException())
+            }
+
+            else -> {
+                cont.resumeWithException(PermissionDeniedException())
+            }
```

Comment:
The conditions result in a lot of duplicated branches
I think this enough

```suggestion
    when (afterStatus) {
        AVAuthorizationStatusAuthorized -> cont.resume(Unit)

        AVAuthorizationStatusDenied ->
            cont.resumeWithException(PermissionDeniedException())

        AVAuthorizationStatusRestricted ->
            cont.resumeWithException(PermissionDeniedPermanentlyException())

        AVAuthorizationStatusNotDetermined ->
            cont.resumeWithException(PermissionNotDeterminedException())

        else ->
            cont.resumeWithException(PermissionDeniedException())
    }